CS3343-Project/src/Project/RR.java

```diff
@@ -13,8 +24,86 @@ public static RR getInstance() {
 
 	@Override
 	public Result schedule(ArrayList<Process> processes) {
-//		Add functionality
-		return null;
+		// storage of loggers for each process
+		HashMap<Integer, ProcessInCPU> loggerMap = new HashMap<Integer, ProcessInCPU>();
+		
+		// create new logger for each session
+		for (int i=0; i<processes.size(); i++) {
+			ProcessInCPU logger = new ProcessInCPU(processes.get(i));
+			loggerMap.put(processes.get(i).getId(), logger);
+		}
+		
+		
+		// main loop
+		for (int curTick=0; curTick < maxLoop; curTick++) {
+			// long term scheduler 
+			for (int i=0; i<processes.size(); i++) {
+				if (processes.get(i).getArrivalTime() == curTick) {
+					readyQueue.add(processes.get(i));
+				}
+				
+			}
+			
+			// keyboard I/O device scheduling
+			if (!blockQueueK.isEmpty()) {
+				Process curIoProcess = blockQueueK.get(0);
+				
+				if (curIoProcess.cur_service_tick >= curIoProcess.getCurServiceTime()) {
+					curIoProcess.proceedToNextService();
+					SystemHelper.moveProcessFrom(blockQueueK, readyQueue);
+					if (!blockQueueK.isEmpty()) {
+						curIoProcess = blockQueueK.get(0);
+					} 
+					else {
+						curIoProcess = null;
+					}
+				}
+				
+				if (curIoProcess != null) {
+					curIoProcess.cur_service_tick++;
+				}
+			}
+			
+			// cpu scheduling
+			if (readyQueue.isEmpty()) {
+				prevProcessId = -1;
+			} 
+			else {
+				boolean loggedWorking = false;
+				Process curProcess = readyQueue.get(0);
+				curProcessId = curProcess.getId();
+				if (curProcessId != prevProcessId) {
+					dispatchedTick = curTick;
+				}
+				
+				curProcess.cur_service_tick++;
+				if (curProcess.cur_service_tick >= curProcess.getCurServiceTime()) {
```

Comment:
Instead of defining state outside of process add a method that returns the state, e.g. curProcess.isCurrentServiceOver() -> returns boolean.